src/pymor/algorithms/timestepping.py

```diff
@@ -10,44 +10,363 @@
+    end_time = None
+    steps = None
+    implicit = False
+    num_values = None
+
+    def __init__(self, initial_time, end_time, num_values=None):
+        assert isinstance(initial_time, Number)
+        assert isinstance(end_time, Number)
+        assert end_time > initial_time
+        assert not num_values or (isinstance(num_values, Number) and num_values > 1)
+        if num_values:
+            self._save_every = (end_time - initial_time) / (num_values - 1)
+        self.__auto_init(locals())
+
+    def bootstrap(self, initial_data, operator, rhs=None, mass=None, mu=None, reserve=False):
+        """Performs required initial work, e.g.
+
+        - setting the initial values for single-step methods or an initial bootstrapping for multi-step methods; or
+        - pre-assembling stationary operators.
+
+        Note: internally calls :meth:_bootstrap, which has to be implemented by every time stepper.
+
+        Parameters
+        ----------
+        initial_data
+            The solution vector at `initial_time`.
+        operator
+            The |Operator| A.
+        rhs
+            The right-hand side F (either |VectorArray| of length 1 or |Operator| with
+            `source.dim == 1`). If `None`, zero right-hand side is assumed.
+        mass
+            The |Operator| M. If `None`, the identity operator is assumed.
+        mu
+            |Parameter values| for which `operator` and `rhs` are to be evaluated. The current
+            time is added to `mu` with key `t`.
+        reserve
+            If True and self.num_values is specified, pre-allocates the returned |VectorArray| for all time steps.
+
+        Returns
+        -------
+        t
+            Simply self.initial_time, to be used in loops as in :meth:`solve`
+        U
+            An empty |VectorArray| to hold the solution trajectory, to be used in loops as in :meth:`solve`.
+        data
+            A dictionary containing information required in :meth:`step` (like previously computed steps), as well
+            as cached data required by the time stepper (such as pre-assembled operators).
+        """
+        if not self.logging_disabled: # explicitly checking if logging is disabled saves some cpu cycles
+            self.logger.debug(f'bootstrapping (mu={mu}) ...')
+
+        assert isinstance(operator, Operator)
+        assert operator.source == operator.range
+
+        rhs = rhs or ZeroOperator(operator.source, NumpyVectorSpace(1))
+        assert isinstance(rhs, (Operator, VectorArray))
+        if isinstance(rhs, Operator):
+            assert rhs.source.dim == 1
+            assert rhs.range == operator.range
+        else:
+            assert rhs in operator.range
+            assert len(rhs) == 1
+            rhs = VectorArrayOperator(rhs)
+
+        mass = mass or IdentityOperator(operator.source)
+        assert isinstance(mass, Operator)
+        assert operator.source == mass.source == mass.range
+        assert not mass.parametric
```

Comment:
I vaguely remember this assertion to be a problem somewhere. Do we actually need it?

Reply:
Not sure, honestly. We do some preassembly in the euler steppers. I can look into it again...

Reply:
removed it